design-proposals/klip-17-sql-union.md

```diff
@@ -0,0 +1,193 @@
+
+**Author**: @big-andy-coates | 
+**Release Target**: TBD | 
+**Status**: In Discussion | 
+**Discussion**: TBD
+
+**tl;dr:** 
+`INSERT INTO` statements allow users to create persistent queries that insert into existing streams.
+However, this is at odds with the target KSQL stream/table model. Switching to SQL's `UNION ALL` 
+operator will allow users to create persistent queries that merge multiple source streams into a 
+new stream. While this may not, in its self, seem to offer much additional functionality to users, 
+it will unlock future enhancements and help clarify user's mental model of KSQL.
+
+## Motivation and background
+
+`INSERT INTO` statements were added to allow many source streams to be merged into a single sink stream.
+Where as a `JOIN` merges the columns from two sources into a row containing columns from both sources, 
+`INSERT INTO` looks to merge at the row level: it can be used to create an output stream that contains all 
+rows of its input streams merged interleaved. For example, consider a system where there are different 
+streams holding orders being placed from different store types, e.g. a stream of `web_orders` and another
+of physical `store_orders`. `INSERT INTO` allowed users to merge these two streams of orders together: 
+
+```sql
+CREATE STREAM all_orders AS SELECT * FROM store_orders;
+INSERT INTO all_orders SELECT * FROM web_orders; 
+```
+
+Resulting in a `all_orders` stream that has a schema that matches both `SELECT`s, and contains all 
+rows from both streams.
+
+However, this approach has several drawbacks.
+
+1. Each additional stream that needs merging into `all_orders` requires a new SQL statement
+and underlying KS topology.
+1. The is no conceptual reason that the first statement should be any different from any other
+statement inserting into `all_orders`, yet the first must be a `CSAS`, while the latter `INSERT INTO`.
+1. It makes it more complicated for a user to drop `all_orders` as they must first terminate all the queries
+that right to it, which may be many. 
+1. It meant there was no longer a (1 - 0..1) relationship between a data source and a persistent query.
+1. It added complexity to KSQL, which now needed to check if there were any `INSERT INTO` queries writing
+to a stream before that stream could be dropped.
+1. It reduced our ability to optimise the topologies and state stores used to build `all_orders`: 
+e.g. if each SELECT writing to `all_orders` was join with some reference table, then each KS topology 
+would have a duplicate state-store holding the materialized reference table. 
+
+The first few drawbacks directly effect our users, making the conceptual model harder to understand 
+and operating KSQL more tiresome, and potentially hurt adoption.  
+
+The later drawbacks are stopping other improvements to KSQL. Including, though not limited to:
+
+* [KLIP-18](klip-18-distributed-metastore .md): Distributed Metastore: which requires the (1 - 0..1) 
```

Comment:
could you add a note on why this is true?

Reply:
Meh... can't you just read KLIP-18? (Once I've raised it)

Reply:
KLIP-18 is now available viewing. Hence I'm not going to add details here.